File: apps/server/src/providers/evolution-client.ts
export interface EvolutionClientOptions {
  baseUrl?: string
  apiKey?: string
  timeoutMs?: number
}

export class EvolutionApiError extends Error {
  constructor(
    message: string,
    readonly status: number,
    readonly payload: unknown,
  ) {
    super(message)
  }
}

export class EvolutionClient {
  readonly baseUrl: string
  private readonly apiKey: string
  private readonly timeoutMs: number

  constructor(options: EvolutionClientOptions = {}) {
    this.baseUrl = (options.baseUrl ?? process.env.EVOLUTION_API_URL ?? 'http://127.0.0.1:8080').replace(/\/$/, '')
    this.apiKey = options.apiKey ?? process.env.EVOLUTION_API_KEY ?? 'local-mvp-evolution-key'
    this.timeoutMs = options.timeoutMs ?? 15_000
  }

  async health(): Promise<unknown> {
    return this.request('GET', '/')
  }

  async fetchInstances(): Promise<unknown> {
    return this.request('GET', '/instance/fetchInstances')
  }

  async createInstance(instanceName: string): Promise<unknown> {
    return this.request('POST', '/instance/create', {
      instanceName,
      integration: 'WHATSAPP-BAILEYS',
      qrcode: true,
    })
  }

  async connectInstance(instanceName: string): Promise<unknown> {
    return this.request('GET', `/instance/connect/${encodeURIComponent(instanceName)}`)
  }

  async connectionState(instanceName: string): Promise<unknown> {
    return this.request('GET', `/instance/connectionState/${encodeURIComponent(instanceName)}`)
  }

  async setWebhook(
    instanceName: string,
    input: { url: string; secret: string },
  ): Promise<unknown> {
    return this.request('POST', `/webhook/set/${encodeURIComponent(instanceName)}`, {
      webhook: {
        enabled: true,
        url: input.url,
        events: [
          'QRCODE_UPDATED',
          'MESSAGES_UPSERT',
          'MESSAGES_UPDATE',
          'SEND_MESSAGE',
          'CONNECTION_UPDATE',
          'CHATS_UPSERT',
          'CHATS_UPDATE',
        ],
        headers: { 'x-control-webhook-secret': input.secret },
        base64: false,
      },
    })
  }

  async logoutInstance(instanceName: string): Promise<unknown> {
    return this.request('DELETE', `/instance/logout/${encodeURIComponent(instanceName)}`)
  }

  async deleteInstance(instanceName: string): Promise<unknown> {
    return this.request('DELETE', `/instance/delete/${encodeURIComponent(instanceName)}`)
  }

  async sendText(instanceName: string, number: string, text: string): Promise<unknown> {
    try {
      return await this.request('POST', `/message/sendText/${encodeURIComponent(instanceName)}`, {
        number,
        text,
        delay: 400,
        linkPreview: true,
      })
    } catch (error) {
      // Older Evolution releases used a nested textMessage object.
      if (!(error instanceof EvolutionApiError) || error.status !== 400) throw error
      return this.request('POST', `/message/sendText/${encodeURIComponent(instanceName)}`, {
        number,
        textMessage: { text },
        delay: 400,
        linkPreview: true,
      })
    }
  }

  async sendMedia(
    instanceName: string,
    input: {
      number: string
      mediatype: 'image' | 'video'
      mimetype: string
      body: Buffer
      caption: string
      fileName: string
    },
  ): Promise<unknown> {
    const form = new FormData()
    form.set('number', input.number)
    form.set('mediatype', input.mediatype)
    form.set('mimetype', input.mimetype)
    form.set('caption', input.caption)
    form.set('fileName', input.fileName)
    form.set('filename', input.fileName)
    form.set(
      'file',
      new Blob([new Uint8Array(input.body)], { type: input.mimetype }),
      input.fileName,
    )
    return this.request(
      'POST',
      `/message/sendMedia/${encodeURIComponent(instanceName)}`,
      form,
      120_000,
    )
  }

  async updateProfilePicture(instanceName: string, picture: string): Promise<unknown> {
    return this.request(
      'POST',
      `/chat/updateProfilePicture/${encodeURIComponent(instanceName)}`,
      { picture },
      60_000,
    )
  }

  async removeProfilePicture(instanceName: string): Promise<unknown> {
    return this.request('DELETE', `/chat/removeProfilePicture/${encodeURIComponent(instanceName)}`)
  }

  async getBase64FromMediaMessage(instanceName: string, message: unknown): Promise<unknown> {
    return this.request(
      'POST',
      `/chat/getBase64FromMediaMessage/${encodeURIComponent(instanceName)}`,
      { message },
      120_000,
    )
  }

  async findChats(instanceName: string): Promise<unknown> {
    return this.request('POST', `/chat/findChats/${encodeURIComponent(instanceName)}`, {
      where: {},
      take: 50,
      skip: 0,
      orderBy: { updatedAt: 'desc' },
    })
  }

  async findMessages(instanceName: string, remoteJid: string): Promise<unknown> {
    return this.request('POST', `/chat/findMessages/${encodeURIComponent(instanceName)}`, {
      where: { key: { remoteJid } },
      take: 50,
      skip: 0,
      orderBy: { messageTimestamp: 'desc' },
    })
  }

  private async request(
    method: string,
    path: string,
    body?: unknown,
    timeoutMs = this.timeoutMs,
  ): Promise<unknown> {
    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), timeoutMs)
    const isMultipart = body instanceof FormData
    try {
      const response = await fetch(`${this.baseUrl}${path}`, {
        method,
        headers: {
          apikey: this.apiKey,
          Accept: 'application/json',
          ...(body === undefined || isMultipart ? {} : { 'Content-Type': 'application/json' }),
        },
        ...(body === undefined ? {} : { body: isMultipart ? body : JSON.stringify(body) }),
        signal: controller.signal,
      })
      const raw = await response.text()
      const payload = raw ? safeJson(raw) : null
      if (!response.ok) {
        const detail = errorDetail(payload) ?? response.statusText
        throw new EvolutionApiError(`Evolution API ${response.status}: ${detail}`, response.status, payload)
      }
      return payload
    } catch (error) {
      if (error instanceof EvolutionApiError) throw error
      if (error instanceof Error && error.name === 'AbortError') {
        throw new Error(`Evolution API 请求超时 (${this.baseUrl})`)
      }
      const detail = error instanceof Error ? error.message : '网络错误'
      throw new Error(`无法连接 Evolution API (${this.baseUrl}): ${detail}`)
    } finally {
      clearTimeout(timeout)
    }
  }
}

function safeJson(raw: string): unknown {
  try {
    return JSON.parse(raw) as unknown
  } catch {
    return raw
  }
}

function errorDetail(payload: unknown): string | null {
  if (typeof payload === 'string') return payload.slice(0, 300)
  if (!isRecord(payload)) return null
  const value = payload.message ?? payload.error ?? payload.response
  if (typeof value === 'string') return value
  return value === undefined ? null : JSON.stringify(value).slice(0, 300)
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null
}

File: apps/server/src/main/java/com/cloudwa/control/provider/EvolutionClient.java
package com.cloudwa.control.provider;

import com.cloudwa.control.config.WaControlProperties;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.web.client.RestClient;
import org.springframework.web.util.UriUtils;

import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

@Component
public class EvolutionClient {
    public final String baseUrl;
    private final String apiKey;
    private final RestClient restClient;
    private final ObjectMapper mapper;

    public EvolutionClient(ObjectMapper mapper, WaControlProperties properties) {
        this.mapper = mapper;
        this.baseUrl = properties.evolution().apiUrl().replaceAll("/$", "");
        this.apiKey = properties.evolution().apiKey();
        this.restClient = RestClient.builder().baseUrl(baseUrl).defaultHeader("apikey", apiKey).defaultHeader("Accept", "application/json").build();
    }

    public Object health() {
        return request("GET", "/", null);
    }

    public Object fetchInstances() {
        return request("GET", "/instance/fetchInstances", null);
    }

    public Object createInstance(String instanceName) {
        return request("POST", "/instance/create", Map.of("instanceName", instanceName, "integration", "WHATSAPP-BAILEYS", "qrcode", true));
    }

    public Object connectInstance(String instanceName) {
        return request("GET", "/instance/connect/" + enc(instanceName), null);
    }

    public Object connectionState(String instanceName) {
        return request("GET", "/instance/connectionState/" + enc(instanceName), null);
    }

    public Object logoutInstance(String instanceName) {
        return request("DELETE", "/instance/logout/" + enc(instanceName), null);
    }

    public Object deleteInstance(String instanceName) {
        return request("DELETE", "/instance/delete/" + enc(instanceName), null);
    }

    public Object findChats(String instanceName) {
        return request("POST", "/chat/findChats/" + enc(instanceName), Map.of("where", Map.of(), "take", 50, "skip", 0, "orderBy", Map.of("updatedAt", "desc")));
    }

    public Object findMessages(String instanceName, String remoteJid) {
        return request("POST", "/chat/findMessages/" + enc(instanceName), Map.of("where", Map.of("key", Map.of("remoteJid", remoteJid)), "take", 50, "skip", 0, "orderBy", Map.of("messageTimestamp", "desc")));
    }

    public Object getBase64FromMediaMessage(String instanceName, Object message) {
        return request("POST", "/chat/getBase64FromMediaMessage/" + enc(instanceName), Map.of("message", message));
    }

    public Object fetchProfilePictureUrl(String instanceName, String number) {
        return request("POST", "/chat/fetchProfilePictureUrl/" + enc(instanceName), Map.of("number", number));
    }

    public Object setWebhook(String instanceName, String url, String secret) {
        return request("POST", "/webhook/set/" + enc(instanceName), Map.of("webhook", Map.of(
                "enabled", true, "url", url,
                "events", List.of("QRCODE_UPDATED", "MESSAGES_UPSERT", "MESSAGES_UPDATE", "SEND_MESSAGE", "CONNECTION_UPDATE", "CHATS_UPSERT", "CHATS_UPDATE"),
                "headers", Map.of("x-control-webhook-secret", secret), "base64", false)));
    }

    public Object sendText(String instanceName, String number, String text) {
        try {
            return request("POST", "/message/sendText/" + enc(instanceName), Map.of("number", number, "text", text, "delay", 400, "linkPreview", true));
        } catch (EvolutionApiException error) {
            if (error.status != 400) throw error;
            return request("POST", "/message/sendText/" + enc(instanceName), Map.of("number", number, "textMessage", Map.of("text", text), "delay", 400, "linkPreview", true));
        }
    }

    public Object sendMedia(String instanceName, String number, String mediaType, String mimeType, byte[] body, String caption, String fileName) {
        LinkedMultiValueMap<String, Object> form = new LinkedMultiValueMap<>();
        form.add("number", number);
        form.add("mediatype", mediaType);
        form.add("mimetype", mimeType);
        form.add("caption", caption);
        form.add("fileName", fileName);
        form.add("filename", fileName);
        form.add("file", new NamedByteArrayResource(body, fileName));
        return restClient.post().uri("/message/sendMedia/" + enc(instanceName)).contentType(MediaType.MULTIPART_FORM_DATA).body(form).retrieve().body(Object.class);
    }

    public Object updateProfilePicture(String instanceName, String picture) {
        return request("POST", "/chat/updateProfilePicture/" + enc(instanceName), Map.of("picture", picture));
    }

    public Object removeProfilePicture(String instanceName) {
        return request("DELETE", "/chat/removeProfilePicture/" + enc(instanceName), null);
    }

    private Object request(String method, String path, Object body) {
        try {
            RestClient.RequestBodySpec spec = restClient.method(HttpMethod.valueOf(method)).uri(path).contentType(MediaType.APPLICATION_JSON);
            String raw = body == null ? spec.retrieve().body(String.class) : spec.body(body).retrieve().body(String.class);
            if (raw == null || raw.isBlank()) return null;
            return mapper.readValue(raw, new TypeReference<Object>() {
            });
        } catch (org.springframework.web.client.HttpStatusCodeException error) {
            throw new EvolutionApiException("Evolution API " + error.getStatusCode().value() + ": " + error.getResponseBodyAsString(), error.getStatusCode().value());
        } catch (EvolutionApiException error) {
            throw error;
        } catch (Exception error) {
            throw new IllegalStateException("无法连接 Evolution API (" + baseUrl + "): " + error.getMessage(), error);
        }
    }

    private static String enc(String value) {
        return UriUtils.encodePathSegment(value, StandardCharsets.UTF_8);
    }

    static class EvolutionApiException extends RuntimeException {
        final int status;

        EvolutionApiException(String message, int status) {
            super(message);
            this.status = status;
        }
    }

    static class NamedByteArrayResource extends ByteArrayResource {
        private final String filename;

        NamedByteArrayResource(byte[] byteArray, String filename) {
            super(byteArray);
            this.filename = filename;
        }

        @Override
        public String getFilename() {
            return filename;
        }
    }
}
